38-allowlist-management.md

```diff
@@ -0,0 +1,162 @@
+
+- [ ] Enhancement issue in release milestone, which links to pull request in [keylime/enhancements]
+- [ ] Core members have approved the issue with the label `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place
+- [ ] User-facing documentation has been created in [keylime/keylime-docs]
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+## Summary
+
+Allowlist management is, obviously, managing allowlists. Currently the allowlist
+is binded with agent records, there are two issues in current design:
+
+* Allowlist reusing: An allowlist can't be shared between agents, so we can't apply
+  the same data to identical environments. By support reusing, we also can provide
+  batch allowlist updates for matching agents.
+* Performance: Since the ima measurements could be very large, it has impact on query
+  efficiency when the allowlist does not participate in the agent processing, e.g., we
+  don't need to load the allowlist for updating agent state.
+
+
+## Motivation
+
+### Goals
+
+The goal is to provide allowlist management API and CLI for allowlist management,
+establish relationship between an allowlist and an agent. The allowlist used by agent
+can be changed when the host of agent is upgraded and the old allowlist doesn't apply.
+
+### Non-Goals
+
+Import allowlist from agent.
+
+## Proposal
+
+The proposal is to:
+
+* Create a new table named ``allowlists`` to hold allowlist data that original hosted
+  in the ``verifiermain`` table, namely: ``tpm_policy``, ``vtpm_policy``, ``and allowlist``.
+* Add a foreign key named ``allowlist_id`` in the ``verifiermain`` table to reference the
+  allowlist data that will be used for the agent.
+* Implements ``/v2/allowlists/`` API for Verifier to support CRD operations. Update an
+  allowlist is not included in the initial proposal but can be implemented if necessary.
+* Enhance ``tenant`` CLI to manage allowlists, when adding agent to the Verifier, user
+  specifies an existing ``allowlist`` name instead of allowlist data to complete the
+  registration.
+
+
+### User Stories (optional)
+
+
+#### Story 1
+
+A user can reuse allowlist by:
+
+* Create an allowlist record by posting policies to the Verifier.
+* Add an agent to the Verifier by providing Agent ID, IP Address and the allowlist name
+  the agent will use.
+* Add another agent running under the same environment to the Verifier with the same allowlist.
+
+#### Story 2
+
+A user can update agent with new allowlist by:
+
+* Create an allowlist record by posting new policies to the Verifier.
+* Update an agent to the Verifier by updating the allowlist name.
+* Agent will use new allowlist to verify system integrety.
+
+### Notes/Constraints/Caveats (optional)
+
+None
+
+### Risks and Mitigations
+
+None
+
+## Design Details
+
+DB Changes:
+
+The ``allowlists`` table will be created as:
+
+```
```

Comment:
Minor nit: GitHub markdown lets you specify the language for blocks. It would make the formatting a bit better to specify it as python